### ISSUE TITLE
Chapter 4v4 comments

### ISSUE THREAD

**Issue by be-marc** [mlr-org/mlr3book#553]
Version 3 of Chapter  Optimization is merged to main. Please add your comments for version 4 here.
PR #552.

New version includes:

* memory management section
* nested resampling example

**Comment by RaphaelS1** [comment]
Hi @be-marc  (CC @berndbischl ) thanks for the update.
But to be honest I really don't understand the "Instructive Example" and I'm not convinced it's useful, some detailed thoughts:

1. I don't understand what the "interleaving half circles" are or what it refers to - so before I even get to the example I am already confused
2. The example is so complicated it can't be run, which goes against everything else in the book
3. The numbers are so close together (1.2%, 1.8%, 1.9%) that I'm not sure this is a convincing argument because there is also just random noise
4. You also don't seem to be comparing like-for-like? There's no tuner defined in the second example

I see the point you're trying to make but unconvinced this is the right approach.

**Comment by RaphaelS1** [comment]
I'd suggest moving memory management to the end of 4.5 or to be honest I think even better as a vignette in mlr3tuning. I think for the vast majority of readers it's too much detail and not relevant and it's more complicated than the other sections - e.g., whilst log transformations is advanced it is at least a feature most people will use, in contrast I've personally never used any of the flags for memory management.

**Comment by be-marc** [comment]
I changed the nested resampling example to a version with a featureless learner. Should be more convincing and works with rendering.

**Comment by sumny** [comment]
Bernd comments:

- [ ] threshold tuning
- [ ] multifidelty to end, separate chapter based on HPO paper and Hyperband blog, skippable clean intro
- [ ] mo to end, skippable, clean intro
- [ ] tuning example: first without log then with log
- [ ] tuning comparison bo, rs, grid 2d sampling behavior
- [ ] rework optimizer table. mostly drop description column, instead brief description of optimizers based on HPO paper
- [ ] recommendations when which tuner, practical stuff

**Comment by mb706** [comment]
I commented the pdf, but then was told to put it up here as bullet points, so people could discuss it. Sorry for the extremely long comment, I didn't expect it to be this much. Each of these contains my comment, followed by a screenshot of the relevant part of the pdf.

[My annotated PDF](https://github.com/mlr-org/mlr3book/files/11042948/chap4_crop.pdf)

### 4. Hyperparameter Optimization
- [ ] 1. This sentence is a bit rough, at the least because these are two main clauses separated by comma, but also because the second half sounds like we want to explain what 'fitting a model' means, when actually we want to explain parameters. Maybe write something like "Parameters are the model coefficients or weights or other information that are determined by the learning algorithm based on the training data". 
![image](https://user-images.githubusercontent.com/15801081/226919863-55181878-2a49-4280-9c9c-4ee76a036455.png)
- [ ] 2. I think this part is a bit cliche. E.g. why talk about neural networks, there are no neural networks in the remainder of the chapter. We are also not talking much about "science" of anything, we are just automatizing things that otherwise one would need to do manually and less reproductively. 
![image](https://user-images.githubusercontent.com/15801081/226922251-b5d40a21-9328-4057-9bd1-3920a2b8a387.png)
- [ ] 3. Consider putting a period after "HPO". Also "follow a numerical black box optimization" should be rephrased 
![image](https://user-images.githubusercontent.com/15801081/226922473-c13e7aa3-9aac-4cd7-87f7-bbab9afb7b05.png)
- [ ] 4. What do we mean with 'robust' here? Low variance? Or does not randomly fail (in which case I wouldn't call it a feature of the resampling method) 
![image](https://user-images.githubusercontent.com/15801081/226922777-e3edd52b-fd77-47c3-94d1-4a47eba50f03.png)
- [ ] 5. I don't like this comparison, since HPO also involves sequential decision making, whereas our benchmarks in mlr3 always use a pre-defined design that does not involve any decisions after the benchmark starts. 
![image](https://user-images.githubusercontent.com/15801081/226923357-a08c287b-e80e-4826-ac4b-cf1e555da619.png)
- [ ] 6. The following figure is described in the text as "naively tun[ing] the number of trees in a random forest", when it shows a perfectly fine grid search on a logarithmic scale. Furthermore, tuning the number of trees is a bit unimaginative, since more trees are just better regarding CE (Probst and Boulesteix even have a paper about this). If we *have* to have an example here, consider using learners that differ on more than one HP, e.g. a few SVMs (which we also use later). But even then we are not showing a major problem of manual tuning: it involves decision making by humans based on results that they see on previous benchmarks, but which may miss some details and are not well reproducible. 
![image](https://user-images.githubusercontent.com/15801081/226925020-b419400e-168a-4eeb-b89b-4791904b1afb.png)
- [ ] 7. As I say above, the example shows grid search, I wouldn't call it more "trial-and-error" than what we would do with a tuner. Also what do we mean with biased / why is it a problem? I would argue that a good HPO algorithm should have a bias towards simpler models. 
![image](https://user-images.githubusercontent.com/15801081/226925696-88098c24-d63a-4a59-b418-23865d752a2b.png)
- [ ] 8. This is a pretty long sentence, maybe use a period after 'levels'. Also remove 'only include'. 
![image](https://user-images.githubusercontent.com/15801081/226926466-67da66c4-c53c-48d8-bbd0-ed32f20ee649.png)
- [ ] 9. I don't feel that strong about this, but maybe don't talk about "higher fidelity" here and instead just call it a full model evaluation.
![image](https://user-images.githubusercontent.com/15801081/226927409-da01fa83-0a58-48aa-80aa-b79405646d7a.png)
- [ ] 10. Maybe say "optimizing" here? 
![image](https://user-images.githubusercontent.com/15801081/226927628-e886960e-25e8-41d6-a926-f0fa4c37d89d.png)
### 4.1.1
- [ ] 11. (a) missing "package". (b) say "specify *that* we want to" (c) the tangent about nu-classification is unnecessary here, leave it out.
![image](https://user-images.githubusercontent.com/15801081/226927933-692782b1-6817-44d2-bb4e-dbb6fbc71f10.png)
- [ ] 12. This a very convoluted example to simply show the hyperparameter setup. Don't we have some data.table printer option that hide superfluous columns automatically? 
![image](https://user-images.githubusercontent.com/15801081/226928925-700cca13-a7f9-4be6-bcf9-d2824e8631e9.png)
- [ ] 13. Missing comma. (I personally also don't like the "infinite resources" phrasing. Depending on what you mean by 'tuning' you'd arguably not even be able to tune (== find the exact optimum of) a single continuous hyperparameter. On the other hand, tuning (== finding a configuration that works better than the default) 16 dimensions indiscriminately may not be a good idea but it is "possible". A better reasoning might be that restricting oneself to certain hyperparameters that are more "tunable" improves results, maybe citing our tunability paper) 
![image](https://user-images.githubusercontent.com/15801081/226929035-b5266651-852b-4269-b924-5e0e368b364b.png)
### 4.1.2
- [ ] 14. (a) reusing the same word too much (b) "which are known as Terminators" --> "which we call Terminators" or "which are represented by Terminator objects"
![image](https://user-images.githubusercontent.com/15801081/226931801-3c0500c4-acea-4863-b20d-0bce53ce740a.png)
- [ ] 15. the account having a budget is usually not the issue here, but instead that cluster jobs have a run time limit. So maybe rephrase this to "can be important on compute clusters where a maximum runtime for a compute job often needs to be specified" 
![image](https://user-images.githubusercontent.com/15801081/226962512-d79a6a95-aa78-4cfc-9f08-49213956b8f6.png)
- [ ] 16. (a) "as if" turns this into a garden path sentence, say "<comma> since if" (b) "no progress is made in" -> "no progress is made for"? also "numer of iterations", not "amount of iterations" (c) "terminated too early" should be rephrased, since it does not explain the concept well. One could think that our implementation of the stagnation terminator is buggy and terminates earlier than it should, given its specification. What we mean here is that it could terminate too aggressively, i.e. that it terminates even though there is some potential left to make substantial progress. (d) "Hyperband" should probably have a reference to somewhere. 
![image](https://user-images.githubusercontent.com/15801081/226963872-d5791d08-cfe1-4965-a90b-a0a227e4002d.png)
- [ ] 17. link does not work in the paper version of the book.
![image](https://user-images.githubusercontent.com/15801081/226963976-1c1b910e-6438-433f-8dc3-858f66d38968.png)
### 4.1.3
- [ ] 18. (a) put the explanation of what a tuning instance is in the beginning, then explain how to construct it. (b) comma before 'or'? 
![image](https://user-images.githubusercontent.com/15801081/226964246-28803b25-5af7-4324-bfd7-ad484ecbfd25.png)
- [ ] 19. 
![image](https://user-images.githubusercontent.com/15801081/226964359-33b4749e-960a-4c2b-b660-daa408efe776.png)
- [ ] 20. 
![image](https://user-images.githubusercontent.com/15801081/226964477-1058f663-a12d-496f-8bd4-73432c6f55eb.png)
### 4.1.4 Tuner
- [ ] 21. 
![image](https://user-images.githubusercontent.com/15801081/226964580-0b5cef4f-71da-4d45-88d4-5a34d2fe6efa.png)
- [ ] 22. our own hyperband is not limited to early stopping, so I would also use a more general term here, or just leave out the 'early stopping' part
![image](https://user-images.githubusercontent.com/15801081/226964858-ea0124b9-872b-4533-85e8-06e65d878921.png)
- [ ] 23. (a) leave out the 'when selecting algorithms' part (b) american english
![image](https://user-images.githubusercontent.com/15801081/226965003-734537fa-92e8-4350-a325-5358461412c5.png)
- [ ] 24. the table 4.2 should mention which tuners come from a package 
![image](https://user-images.githubusercontent.com/15801081/226965159-d8ba7c28-c403-4855-a6d1-5e4dd4ef713b.png)
- [ ] 25. (a) mention that the batch_size is only really relevant when doing parallelization (ref parallelization chapter) and will also set the interval in which terminators are checked (i.e. batch size 5 means that multiple of 5 configs are evaluated, even if there is an evals-terminator with e.g. n_evals = 6). (b) I think it is a bad idea to have `batch_size` equal to the `resolution` here, since one may think that these are connected (i.e. that one should usually set the batch size to the resolution value). I would just leave it out of this example, or set it to the `future::` detected number of workers.
- [ ] 26. maybe write 'and' here? 
![image](https://user-images.githubusercontent.com/15801081/226966074-af1168af-ff8e-48ef-bb2b-d3b7858fcb76.png)
- [ ] 27. This sounds to me like *one* model is fitted to the *entire task* for each config; instead for each config we fit multiple models on task subsets (resampling). Maybe write "... set of hyperparameter values that is evaluated on the given task using resampling"
![image](https://user-images.githubusercontent.com/15801081/226966456-93ee07e3-018e-400a-9789-b9aed93f8a16.png)
- [ ] 28. (a) "just like learners, tuners also have parameters" sounds weird after we have stressed before that the 'parameters' of a learner are what the algorithm constructs itself, not the configuration options for the user. Maybe write that "similar to learners, which have hyperparameters, one can configure tuners through what we call 'control parameters'"?  (b) controls -> control (plural) (c) "rarely need to be changed" is quite opinionated here (do we have a tune() call in the book where we don't change a tuner's config parameters?); instead write that many of them perform well enough in their default setting, but changing them can still improve tuning performance substantially.
![image](https://user-images.githubusercontent.com/15801081/226966773-1b5cc163-2326-43a7-90cc-d6f43e325c58.png)
### 4.1.5
- [ ] 29. I wouldn't say "trigger", maybe "run", "start", etc. the tuning process? Also using the exclamation mark sounds a bit too exciting imho 
![image](https://user-images.githubusercontent.com/15801081/226968278-8a8aa158-af8c-456a-8b29-4a6fd2e08d42.png)
- [ ] 30. (a) 'transformed values' should have a reference to what we mean with transformation right there, not in the next sentence. (b) the statement 'none in this example' is in conflict with the value of `learner_param_vals`, which is clearly not empty.
![image](https://user-images.githubusercontent.com/15801081/226967913-6968de82-f88c-4f19-a93c-f73fde8a1879.png)
### 4.1.6
- [ ] 31. (a) Don't write "more control" here, it is redundant with the 'less control' later in the sentence, and at this point we haven't mentioned anything that we are comparing it to. (b) "sugar function" -> "helper function" 
![image](https://user-images.githubusercontent.com/15801081/226968422-0bc9e395-9726-438d-8e82-61b2834c8412.png)
- [ ] 32. In another chapter I was told not to show an example where an error is triggered when the user does something wrong, showcasing how mlr3 prevents certain undesirable things from happening. There was no danger of the user misinterpreting this and using this in his code, since it would also trigger an error for him. Here, however, we are giving an example of how one should clearly NOT tune on this kind of search space, and there are NO guardrails in mlr3 that prevent this from happening. We should absolutely not have this code in the book, not even with some hints further down that one would not do it like this, since people are still going to copy this example code and then wonder why tidymodels tunes the SVM so much better. If it has to be the SVM with these parameters, use `logscale = TRUE`, even if it uses trafos right away. Everybody should use `logscale = TRUE` all the time, so it is good to get them used to it. 
![image](https://user-images.githubusercontent.com/15801081/226970061-46f952d7-40a4-4516-9189-6849b30a37d7.png)
### 4.1.7
- [ ] 33. Tone is a bit too colloquial here 
![image](https://user-images.githubusercontent.com/15801081/226970334-1b8fbfe4-b0cd-4be0-a944-6bccc23b363a.png)
- [ ] 34. This can be seen from the printout, no need to write this here
![image](https://user-images.githubusercontent.com/15801081/226970444-ac6f8c54-861d-47a5-a8ae-99cdfbe4e516.png)
- [ ] 35. Comma should be a period
![image](https://user-images.githubusercontent.com/15801081/226970512-4906534a-42dd-457d-8585-8474d705717d.png)
- [ ] 36. Period before "however", also comma afterwards. 
![image](https://user-images.githubusercontent.com/15801081/226970759-8b563d7e-ef99-4286-acc7-793ce04914bc.png)
- [ ] 37. (a) the "timestamp" does not show how long each evaluation took. (b) also mention warnings. (c) in -> while
![image](https://user-images.githubusercontent.com/15801081/226971017-110d7c35-ca49-42cd-beea-e63f86c521ca.png)
- [ ] 38. 
![image](https://user-images.githubusercontent.com/15801081/226971220-776fc134-334e-40f4-b350-3cb17208446b.png)
- [ ] 39. Same as above: data.table hides many rows, we don't necessarily see that. 
![image](https://user-images.githubusercontent.com/15801081/226971343-b26b7ae4-dd86-489f-8518-86a20d58b36d.png)
- [ ] 40. (a) view -> access (b) with -> "through"? "as"? 
![image](https://user-images.githubusercontent.com/15801081/226971700-884843e3-6095-4d76-b7bb-6d73a96b0216.png)
- [ ] 41. comma? 
![image](https://user-images.githubusercontent.com/15801081/226971764-2e9e261b-d985-4a4f-9982-730df4ab0f8e.png)
- [ ] 42. general comment, the atoplot here fits a ranger model and plots that. This makes it look like the area around (0.1, 0.1) is better than the area around (25000, 0.3), althought the points themselves show that (0.1, 0.1) has worse performance.
![image](https://user-images.githubusercontent.com/15801081/226972300-d5326080-94e3-43a6-938d-1eab050c4ac0.png)
### 4.1.8 Using a tuned model
- [ ] 43. (a) with -> to (b) configurations -> configuration (singular, since we only have the result configuration) (c) setting 'id' is not necessary here and distracts 
![image](https://user-images.githubusercontent.com/15801081/226972534-55697dec-b7e7-4b26-b42f-9f89670edb67.png)
- [ ] 44. (a) 'as the model's performance' -> 'as an estimate of the model's performance' (b) 'therefore' repeated
![image](https://user-images.githubusercontent.com/15801081/226972797-0c24c00f-c9e3-448c-8c16-f278ac42a6cf.png)
### 4.2 Multi-Objective Tuning
- [ ] 45. insert commas here?
![image](https://user-images.githubusercontent.com/15801081/226973077-82bf6925-1692-4c68-9b16-d51e40022615.png)
- [ ] 46. In a more -> as a more 
![image](https://user-images.githubusercontent.com/15801081/226973307-1844af14-65a9-4e52-8cd6-7510df5751aa.png)
- [ ] 47. Absolute winner here. We actually found ourselves writing a book containing the string "there is clearly an ethical argument to [...], however" 
![image](https://user-images.githubusercontent.com/15801081/226973457-a7b00f90-1152-41ec-9af8-cd825e072b1e.png)
- [ ] 48. 
![image](https://user-images.githubusercontent.com/15801081/226974032-39c62d3c-83fd-4737-af53-c4c47b8c3246.png)
- [ ] 49. period instead of comma? 
![image](https://user-images.githubusercontent.com/15801081/226974217-00e9ef76-f9e3-4565-bbbf-06e6b1addbff.png)
- [ ] 50. (a) comma after 'Pareto' (b) refer to the figure, it does not come right afterwards.
![image](https://user-images.githubusercontent.com/15801081/226974405-0868540e-c813-4a31-ab53-6f387cc59755.png)
- [ ] 51. The table contains multiple entries with the same objective values, maybe we don't want to list all of this
![image](https://user-images.githubusercontent.com/15801081/226974593-eb1b573b-cc00-4e80-a724-27ba4a819385.png)
### 4.3. Automated Tuning with AutoTuner
- [ ] 52. "powerful" is a weird word choice, maybe write 'useful'? 
![image](https://user-images.githubusercontent.com/15801081/226974748-bcaa00f8-fba9-460c-9dd6-cc828789b20d.png)
- [ ] 53. I would phrase this as "without the need to construct a new learner with the tuned configuration"
![image](https://user-images.githubusercontent.com/15801081/226974995-e4e87199-1c5b-48e5-8ce0-a575ad71732b.png)
- [ ] 54. I will complain about this every time I see it
![image](https://user-images.githubusercontent.com/15801081/226975048-791a4a32-cdec-48cd-b8dd-6e3a945a2ebf.png)
- [ ] 55. (a) remove comma (b) "tunes the learner's hyperparameters before fitting the model" 
![image](https://user-images.githubusercontent.com/15801081/226975600-31d246ab-19bf-457d-b297-2f003bd1e231.png)
- [ ] 56. no need for the reference here
![image](https://user-images.githubusercontent.com/15801081/226975757-2bcbecd8-644e-4f90-bed1-0ced511fc927.png)
### 4.4 Nested Resampling
- [ ] 57. (a) don't call this a 'layer', in ML context this too often means neural network layers (b) be more specific: this only introduces a bias when estimating performance, the tuning itself is not biased 
![image](https://user-images.githubusercontent.com/15801081/226976053-9ff27452-df55-47fd-90c2-68cb829bd21b.png)
- [ ] 58. (a) performance of the tuned / resulting model? What we are really doing is estimating the performance of the tuning procedure, maybe good to be more precise here (b) american english 
![image](https://user-images.githubusercontent.com/15801081/226976713-db1a503f-0f21-425e-8e04-955c8c35b2fa.png)
- [ ] 59. (a) a bit bloated (b) remove comma (c) a bit too colloquial
![image](https://user-images.githubusercontent.com/15801081/226977146-2b5ea1c3-9926-41c6-be74-14bd878ee109.png)
- [ ] 60. mention that this is an example 
![image](https://user-images.githubusercontent.com/15801081/226977279-c0960bd1-1816-4cc0-a961-4c6a905052c7.png)
- [ ] 61. (a) write 'outer' training data (b) comma before 'instantiate' (not in the image)
![image](https://user-images.githubusercontent.com/15801081/226977445-b8bf921a-e585-460c-ac29-3bc4005e059f.png)
- [ ] 62. insert 'on the outer training set,' 
![image](https://user-images.githubusercontent.com/15801081/226977552-2daba0a0-472d-43ed-9993-31ff6428bd40.png)
- [ ] 63. Be less colloquail, probably just remove this sentence.
![image](https://user-images.githubusercontent.com/15801081/226977628-26edab27-71d2-4b8b-ab18-d176ba95c9dc.png)
- [ ] 64. comma
![image](https://user-images.githubusercontent.com/15801081/226977691-cc0cacae-d98f-411e-b4c3-53143253a64f.png)
- [ ] 65. logscale
![image](https://user-images.githubusercontent.com/15801081/226977738-67939a68-e6c1-4203-92ff-e414e17b48ad.png)
- [ ] 66. (a) "Note that" (b) "to make it possible to investigate", or "so that the inner tuning can be investigated later" (c) comma (d) just remove 'controlling'
![image](https://user-images.githubusercontent.com/15801081/226978039-c32d5db5-e137-4126-b6e2-84e1430dd76a.png)
- [ ] 67. comma 
![image](https://user-images.githubusercontent.com/15801081/226978153-ab35157b-d957-4ad5-8e78-eb48ed06a9d4.png)
- [ ] 68. (a) maybe remove comma? (b) remove 'greatly' (c) stress "outer" resampling iterations (d) and (e) are just wrong, the problem is not stability / convergence (just like irace, random search also "converges to an optimal configuration" in the limit), but that different folds may have different actual optima, and that these may be different from the optimum wen optimized on the entire training data. 
![image](https://user-images.githubusercontent.com/15801081/226979557-4d1f18c1-4808-4600-94f3-85ab1a85c21d.png)
- [ ] 69. performance on the inner resampling folds appears to be better in this example
![image](https://user-images.githubusercontent.com/15801081/226979688-73f131a6-504b-48d4-9570-c74ac8f448e3.png)
- [ ] 70. (a) *estimated* performance (b) remove "always", no need to get so serious
![image](https://user-images.githubusercontent.com/15801081/226979994-025902d3-a677-4b6f-a0d3-ab45fd63bcbb.png)
- [ ] 71. Weird tangent to go on here; just remove this part, or mention the more relevant other concept of "bias" instead, which would be the bias of model parameters
![image](https://user-images.githubusercontent.com/15801081/226980597-b5858443-e004-4212-99fd-dedc0709b445.png)
- [ ] 72. Why construct a "simple example", just use the numbers from above
![image](https://user-images.githubusercontent.com/15801081/226981410-cb3a845a-81ba-4f7f-bb4b-6cc3d9e5de79.png)
### 4.4.3
- [ ] 73. this subchapter is superfluous and should be removed
![image](https://user-images.githubusercontent.com/15801081/226981541-3ce4bfbf-813c-431c-b7de-cc99082ea6f5.png)
- [ ] 74. comma 
![image](https://user-images.githubusercontent.com/15801081/226981642-676d61e4-d806-42a8-96e0-f2f82a53696c.png)
- [ ] 75. (a) don't call this "buffer", it has a different meaning in software, which is confusing here. (b) remove comma?
![image](https://user-images.githubusercontent.com/15801081/226981826-52dfdf02-fb88-4f72-ae24-e274b1135d7f.png)
- [ ] 76. (a) remove 'in the code above', it is clear from context (b) "accept" -> "ignore" (c) 'manually debug' skips a step in the reasoning chain here. Explain that we are "happy to ignore errors in training, but not doring prediction, and want the tuning to be aborted should an error happen during prediction" or something like this (d) explain that `evaluate` uses R's own error handling mechanism and runs the learner in the same R process that the tuner runs in (e) mention callr's main advantage, that it can handle learners that crash the R session (f) comma before 'which' here (g) remove 'in mlr3 you can do this', or at least turn the comma into a period. (g) Turn the last comma into a period.
![image](https://user-images.githubusercontent.com/15801081/226982964-7502876a-0cb5-4d34-8090-1c2d806a088f.png)
- [ ] 77. (a) "predict the mean of the *target*" (b) comma before 'or', (c) remove 'just'. Possibly also mention 'regression' and 'classification' somewhere in this sentence.
![image](https://user-images.githubusercontent.com/15801081/226983248-494b7481-e632-4168-b89e-2a3eb0fb99aa.png)
- [ ] 78. This sentence at first made me think that a lot of working memory is used by the algorithm, when really we are concerned about storage of results here. Maybe write "Storing the result of a large tuning experiment can require a lot of memory" / "storage space"? 
![image](https://user-images.githubusercontent.com/15801081/226983630-767a26e9-ba24-4845-bb49-42969806a965.png)
### 4.5.2
- [ ] 79. I think it is not clear enough that "store_models = FALSE" is the default here, and that, in fact, nothing needs to be done here.
![image](https://user-images.githubusercontent.com/15801081/226983821-2565b616-7ce4-4b83-893e-26104257fefd.png)
- [ ] 80. (a) comma after 'results' (b) mention that we are talking about getting the performance measures offline; we would still be able to use the recorded hyperparameters and do another resampling to estimate other measures
![image](https://user-images.githubusercontent.com/15801081/226984156-2f0f7ecb-3a27-4a75-8ac4-a3d9cbf4e967.png)
- [ ] 81. 
![image](https://user-images.githubusercontent.com/15801081/226984249-c020b44f-c384-4d8c-8146-8bfc5fc23d0f.png)
- [ ] 82. I don't understand what this is supposed to tell me. In particular the 'store_tuning_instance' is set twice?
![image](https://user-images.githubusercontent.com/15801081/226984383-85c081ac-fa3e-4498-8967-60c4c9bb4718.png)
- [ ] 83. We mention 'store_models = FALSE' in the first and fourth paragraph in this subchapter, maybe mention both aspects in the same paragraph?
![image](https://user-images.githubusercontent.com/15801081/226984551-f5fb1587-038f-492a-897a-9d29275ade69.png)
## 4.6
This chapter seems to be in a noticeably better shape than the rest
- [ ] 84. "can be constructed" 
![image](https://user-images.githubusercontent.com/15801081/226984897-a5c14d16-8c76-4982-8405-83bf42c8cd8c.png)
- [ ] 85. I don't think this detail is important here
![image](https://user-images.githubusercontent.com/15801081/226984976-3ed9ed1e-a505-4123-8aa6-e10d28d4c42e.png)
- [ ] 86. Forward-reference to the pipelines chapter? 
![image](https://user-images.githubusercontent.com/15801081/226985068-df57e998-1afc-4c59-81b3-d407e6231ae5.png)
- [ ] 87. I think this can be skipped, it does not add much
![image](https://user-images.githubusercontent.com/15801081/226985159-2e4a04b5-b9b1-4b79-9d53-f637195a1a7f.png)
- [ ] 88. clearly missing the trafo here; make it `p_dbl(0, 1)`
![image](https://user-images.githubusercontent.com/15801081/226985296-2d7dc78e-1d10-42a7-a35c-c1b89504ecc6.png)
- [ ] 89. "an evaluation" -> "a configuration" or "a hyperparameter configuration"
![image](https://user-images.githubusercontent.com/15801081/226985415-c78637ed-14ed-4e15-94af-804ac9889b66.png)
### 4.6.2 Creating Search Spaces from Learners
- [ ] 90. "allows" -> "makes it possible"?
![image](https://user-images.githubusercontent.com/15801081/226985654-b3d03773-584b-4c63-86ca-6aeb84f09fd0.png)
- [ ] 91. missing space
![image](https://user-images.githubusercontent.com/15801081/226985708-d2bdcde8-421e-4028-900b-0a5b98d1c8c3.png)
### 4.6.3 Search Space Collection
- [ ] 92. Maybe use a more descriptive title, e.g. "predefined" or "recommended" search spaces?
![image](https://user-images.githubusercontent.com/15801081/226985957-5fef53e2-b4a6-43d9-86f9-bb101f7da994.png)
- [ ] 93. Also the chapter has too much output
### 4.7 Conclusion
- [ ] Also mention search spaces here?
![image](https://user-images.githubusercontent.com/15801081/226986168-f33a0d0d-dd68-4332-9e6e-13ef56458ba4.png)
- [ ] online documentation?
![image](https://user-images.githubusercontent.com/15801081/226986241-48fb9d5d-921a-4bf0-a62a-9efd86274cb1.png)

**Comment by sumny** [comment]
currently I don't like: intro is after all these reiterations very very basic and not even the concept of GE is mentioned anymore still we use the term sometimes; maybe see how this is introduced in chapter 3 and resolve this

**Comment by berndbischl** [comment]
i went thru ALL of martins comments with marc. i am ok with ALL of marcs and lennarts changes and had 2 additional "30 sec" comments. pls merge and give to lars
